Use provided `GITHUB_*` URL variables

rkarp opened the pull request:
Here's the initial implementation of the GitHub URLs part of #86.

There is also the option of adding these variables to the CLI parameters and using the `env` clap attribute to read the defaults, but I can't think of a use case where someone would realistically want to run the action on a self-hosted GitHub instance but delete images from central GitHub or vice versa.

sondrelg reviewed (commented):
This looks like a great starting point! 

I'm not sure we need to introduce a new struct for these. Could we make something like this change to the `Urls` struct?

```rust
impl Urls {
    fn load_base_urls_from_env() -> Result<(String, String)> {
        let server_url = match env::var("GITHUB_SERVER_URL") {
            Ok(value) => Ok(value),
            Err(error) => {
                if error == VarError::NotPresent {
                    Ok("https://github.com".to_string())
                } else {
                    Err(error)
                }
            }
        };
        let api_url = match env::var("GITHUB_API_URL") {
            Ok(value) => Ok(value),
            Err(error) => {
                if error == VarError::NotPresent {
                    Ok("https://api.github.com".to_string())
                } else {
                    Err(error)
                }
            }
        };
        
        Ok((server_url?, api_url?))
    }
    
    pub fn from_account(account: &Account) -> Self {
        let (mut github_base_url, mut api_base_url) = Self::load_base_urls_from_env().expect("Failed to load Github URLs from the environment");
```

open to discuss pros/cons of both approaches - to me, this would be easier to follow, since all the relevant code would be contained in one place. Perhaps there are more idiomatic ways to do it.

Would also be great to add some doc comments linking to the Github documentation you referenced in the issue, and we should probably add a test to prevent this from breaking in the future. I could also do these if you prefer  ☺️

rkarp commented:
Something similar to your example was my first thought as well, but then I noticed that the `urls` module would be better off staying without IO code (so no environment variable reading) and thus cleanly unit-testable. So reading all config values (args and env) should ideally be outside the specialized, pure modules. `main` seemed like a good interim location.

The `UrlBases` struct is not my favorite permanent solution either. The URL fields could instead be added to the `Input` struct and loaded from CLI parameters / env / default, which clap would support nicely. But that seemed to significant of a change for now. I'm beginning to like this variant more, though, since it would allow users to use the CLI outside GitHub Actions and still target self-hosted GitHub packages.

Another idea might be to introduce a new `Config` struct, which would contain `Input` plus all extra values (like the URLs) that should only be read from environment variables. That, too, would need a new module or a rework / rename of the `args` module. It may be a better idea to instead just decide on letting every parameter be set by CLI as well and use `Input` only.

With both variants, one could then pass either `Input` or `Config` (instead of just `Account`) to `Urls::new`, to avoid having to add more parameters over time if you want to make everything more configurable in the future.

What do you think about these options?

sondrelg commented:
> The URL fields could instead be added to the Input struct and loaded from CLI parameters / env / default, which clap would support nicely. But that seemed to significant of a change for now. I'm beginning to like this variant more, though, since it would allow users to use the CLI outside GitHub Actions and still target self-hosted GitHub packages.

What would this look like? I think I would prefer this 👍

rkarp commented:
New version is up.

`url`'s handling of slashes is in some cases surprising (see servo/rust-url#808 and servo/rust-url#713), but I liked the idea of getting the URL format immediately checked by clap, so I went with `Url` instead of `String`.

Test coverage seems decent already. With docs I wasn't sure what you would prefer, so I might leave the rest (Readme, possible doc comments) to you.

JonasKs commented on src/client/client.rs:
```diff
@@ -571,43 +576,51 @@ mod tests {
         assert_eq!(
             urls.list_package_versions_url("foo").unwrap().as_str(),
-            "https://api.github.com/orgs/acme/packages/container/foo/versions?per_page=100"
+            Input::DEFAULT_GITHUB_API_URL.to_string() + "/orgs/acme/packages/container/foo/versions?per_page=100"
```
Wouldn't this be  
```
https://api.github.com//orgs/acme/packages/container/foo/versions?per_page=100
                      ^ two `//`'s here
```
considering you add a `/` in the `Url::parse((url_str.to_string() + "/").as_str())`?

JonasKs replied:
Oh, I now see that you removed the prefix from the URLs in the code, and just read your previous comment about slashes. So this is probably not an issue, then?

rkarp replied:
Yes, it's not an issue, the tests would have failed otherwise. Adding a slash is only for the conversion to `Url`, since the `url` library adds them (but only on an empty path).

sondrelg commented:
I made a few changes on top of your branch in https://github.com/snok/container-retention-policy/pull/91 @rkarp 👍 I'm not good enough at git to push to this PR without a few minutes more googling than I have available right now 😅 

It:

- Replaces `Url::try_from` with `Url::parse`, since `try_from` just calls `parse` under the hood
- Adds logging when non-default URLs are loaded
- Adds a test for the logging
- Makes the default urls global (personal preference)

and I tested the code with and without the `try_parse_url` code, and it worked for both, so I removed that.

Want to take a look and see if you agree with the changes made?

rkarp commented:
The changes look good, except removing `try_parse_url`. That one is needed, otherwise the URL generation would break if the user provides a URL with a path, but no `/` at the end, like `https://foo.com/bar`. Additionally, there's the issue of [`Url.join`](https://docs.rs/url/2.5.2/url/struct.Url.html#method.join) overwriting the final path component if it doesn't end with `/`, so that's a good reason to make sure that URL Strings that refer to paths should always end with `/` before converting them to `Url`.

There are also some small inconsistencies in the refactoring commits that I could clean up (I'm quite good at git). I could remove your last commit and squash the refactoring commits, which would be the easiest way to clean everything up. Is that OK with you?

rkarp commented:
I've fixed your commits and also added a correction for logging. This wouldn't have worked correctly since `Url::parse` adds a `/` on an empty path, as mentioned:

```rust
input.github_server_url.as_str() != DEFAULT_GITHUB_SERVER_URL
```

sondrelg commented:
Sounds good @rkarp!

rkarp commented:
Alright, done.

sondrelg commented:
Cheers! Just a quick note to say I've left for holidays, so will get this merged and released soon, but it might take a few days. Thanks again for the work here @rkarp 👍

rkarp commented:
Sure, there's no hurry!
